metrics/metrics.go

```diff
@@ -389,27 +389,79 @@ func GroupHandler() http.Handler {
 	return promhttp.HandlerFor(GroupMetrics, promhttp.HandlerOpts{Registry: GroupMetrics})
 }
 
-// lazyPeerHandler is a structure that defers learning who current
-// group members are until an http request is received.
+// lazyPeerHandler is a structure that defers learning which peers this node is connected
+// to until an http request is received for a specific peer. It handles all peers that
+// this node is connected to regardless of which group they are a part of.
 type lazyPeerHandler struct {
-	peerHandler PeerHandler
+	metricsHandlers []Handler
+	// handlerCache is a cache of peer address -> handler
+	// TODO Do we need to evict from cache at some point? The only case when it should be
+	//      invalidated is if a peer leaves e.g. during a Reshare
+	handlerCache sync.Map
+}
+
+// newLazyPeerHandler creates a new lazyPeerHandler from a slice of GroupHandlers
+func newLazyPeerHandler(metricsHandlers []Handler) *lazyPeerHandler {
+	return &lazyPeerHandler{
+		metricsHandlers,
+		sync.Map{},
+	}
 }
 
+// handlerForPeer returns the http.Handler associated with a given peer address.
+// Metrics are group-agnostic. Therefore, we just need the Handler for a group that the
+// peer in question has joined, regardless of the group. If a peer belongs to multiple
+// groups, it will return the handler associated with the group that appears first in
+// the metricsHandlers whose beacon the peer has joined.
+//
+// If the peer is not found, it will return a nil handler and a nil error.
+// If there is any other error it will return false and the given error.
+func (l *lazyPeerHandler) handlerForPeer(addr string) (http.Handler, error) {
+	h, found := l.handlerCache.Load(addr)
+	if found {
+		return h.(http.Handler), nil
+	}
+
+	var err error
+
+	for _, hdl := range l.metricsHandlers {
+		h, err = hdl(addr)
+		if err != nil {
+			//nolint:errorlint
+			switch err.(type) {
+			case *common.NotPartOfGroupError:
+				// The target addr hasn't joined this beacon. Try the next one.
+				continue
+			default:
+				return nil, err
+			}
+		}
+
+		l.handlerCache.Store(addr, h)
+		return h.(http.Handler), nil
+	}
+
+	return nil, nil
+}
+
+// ServeHTTP serves the metrics for the peer whose address is given in the URI. It assumes that the
+// URI is in the form /peer/<peer_address>
 func (l *lazyPeerHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	addr := strings.Replace(r.URL.Path, "/peer/", "", 1)
 	if index := strings.Index(addr, "/"); index != -1 {
 		addr = addr[:index]
 	}
 
-	handlers, err := l.peerHandler(r.Context())
+	handler, err := l.handlerForPeer(addr)
+
 	if err != nil {
-		log.DefaultLogger().Warnw("", "metrics", "failed to get peer handlers", "err", err)
+		log.DefaultLogger().Warnw("", "metrics", "failed to get handler for peer", "addr", addr, "err", err)
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
-	handler, ok := handlers[addr]
-	if !ok {
+	if handler == nil {
```

Comment:
This can never happen when `err == nil`, can it?

Reply:
See https://github.com/drand/drand/pull/1017/files#diff-d715f391f2ffa038c3167bfea1adf741c1f5e39f08349fbbacab780a594146a4R444. `nil, nil` is returned when there is no handler. I'll use the same pattern as for `ErrNoGroup` to make it clearer.